Check constraint kills server

**BUG REPORT**

- What did you do?
```sql
create database test;
set database = test;
create table test (i int check (i > 10));
insert into test values ('11'); -- deliberately input a string 
```
- What did you expect to see?
```
pq: failed to satisfy CHECK constraint (i > 10)
```
- What did you see instead?
```
driver: bad connection
connection lost; opening new connection and resetting session parameters...
```
The docker instance was killed and needed to be restarted manually.
After logging in, trying to drop the table
```sql
drop table test.test;
--pq: table "test" is being dropped
```
It got stuck and took ~ 10 mins for the table to be finally dropped.

Host: Windows 10 Pro
Tested on cockroachdb docker beta-20170413-dirty and alpha-18537-g00185f3


Possibly related to #14513 

One liner for reproducibility
```
./cockroach sql -e "create database test; set database = test; create table test (i int check (i > 10)); insert into test values ('11');"
```

Here are the 2 logs that I found from a brand new docker instance:
```
root@1c2a9a631f5f:/cockroach# cat cockroach-data/logs/cockroach.log
I170417 01:34:44.764377 1 util/log/clog.go:1007  [config] file created at: 2017/04/17 01:34:44
I170417 01:34:44.764377 1 util/log/clog.go:1007  [config] running on machine: 1c2a9a631f5f
I170417 01:34:44.764377 1 util/log/clog.go:1007  [config] binary: CockroachDB CCL beta-20170413-dirty (linux amd64, built 2017/04/13 21:06:17, go1.8.1)
I170417 01:34:44.764377 1 util/log/clog.go:1007  [config] arguments: [/cockroach/cockroach start --insecure]
I170417 01:34:44.764377 1 util/log/clog.go:1007  line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg utf8=✓
I170417 01:34:44.764376 1 cli/start.go:390  CockroachDB CCL beta-20170413-dirty (linux amd64, built 2017/04/13 21:06:17, go1.8.1)
I170417 01:34:44.865177 10 cli/start.go:427  starting cockroach node
W170417 01:34:44.865236 10 server/server.go:158  [n?] running in insecure mode, this is strongly discouraged. See --insecure.
I170417 01:34:44.868485 10 storage/engine/rocksdb.go:379  opening rocksdb instance at "/cockroach/cockroach-data"I170417 01:34:45.243848 10 server/config.go:403  1 storage engine initialized
I170417 01:34:45.244177 10 server/server.go:651  [n?] sleeping for 123.9792ms to guarantee HLC monotonicity
I170417 01:34:45.373085 10 storage/store.go:1256  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170417 01:34:45.373124 10 server/node.go:456  [n1] initialized store [n1,s1]: {Capacity:63143981056 Available:56824078336 RangeCount:12 LeaseCount:0}
I170417 01:34:45.373146 10 server/node.go:347  [n1] node ID 1 initialized
I170417 01:34:45.373213 10 gossip/gossip.go:307  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"1c2a9a631f5f:26257" > attrs:<> locality:<>
I170417 01:34:45.373297 10 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
I170417 01:34:45.373398 10 server/node.go:606  [n1] connecting to gossip network to verify cluster ID...
I170417 01:34:45.373418 10 server/node.go:631  [n1] node connected via gossip and verified as part of cluster "2aa6cf90-c607-4d56-9d71-df535279c836"
I170417 01:34:45.373442 10 server/node.go:394  [n1] node=1: started with [[]=/cockroach/cockroach-data] engine(s) and attributes []
I170417 01:34:45.373493 10 sql/executor.go:326  [n1] creating distSQLPlanner with address {tcp 1c2a9a631f5f:26257}
W170417 01:34:45.374245 42 storage/store.go:1343  [n1,s1,r5/1:/Table/{0-11},@c420162c00] could not gossip system config: [NotLeaseHolderError] range 5: replica {1 1 1} not lease holder; lease holder unknown
I170417 01:34:45.377804 10 server/server.go:705  [n1] starting http server at 1c2a9a631f5f:8080
I170417 01:34:45.377819 10 server/server.go:706  [n1] starting grpc/postgres server at 1c2a9a631f5f:26257
I170417 01:34:45.377833 10 server/server.go:707  [n1] advertising CockroachDB node at 1c2a9a631f5f:26257
I170417 01:34:45.378447 10 server/server.go:815  [n1] done ensuring all necessary migrations have run
I170417 01:34:45.378461 10 server/server.go:817  [n1] serving sql connections
I170417 01:34:45.378525 10 cli/start.go:497  Startup message:
CockroachDB node starting at 2017-04-17 01:34:45.3784754 +0000 UTC
build:      CCL beta-20170413-dirty @ 2017/04/13 21:06:17 (go1.8.1)
admin:      http://1c2a9a631f5f:8080
sql:        postgresql://root@1c2a9a631f5f:26257?sslmode=disable
logs:       /cockroach/cockroach-data/logs
store[0]:   path=/cockroach/cockroach-data
status:     restarted pre-existing node
clusterID:  2aa6cf90-c607-4d56-9d71-df535279c836
nodeID:     1
I170417 01:34:45.393216 91 sql/event_log.go:101  [n1] Event: "node_restart", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:1c2a9a631f5f:26257} Attrs: Locality:} ClusterID:2aa6cf90-c607-4d56-9d71-df535279c836 StartedAt:1492392885373421000 LastUp:1492392869248226500}
I170417 01:34:55.377592 46 storage/store.go:3966  [n1,s1] sstables (read amplification = 1):
6 [ 120K 1 ]: 120K
I170417 01:34:55.377664 48 server/status/runtime.go:227  [n1] runtime stats: 50 MiB RSS, 62 goroutines, 8.5 MiB/3.6 MiB/17 MiB GO alloc/idle/total, 5.2 MiB/8.6 MiB CGO alloc/total, 0.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (5x)
I170417 01:35:05.377667 48 server/status/runtime.go:227  [n1] runtime stats: 52 MiB RSS, 62 goroutines, 9.4 MiB/3.1 MiB/18 MiB GO alloc/idle/total, 5.1 MiB/8.7 MiB CGO alloc/total, 83.20cgo/sec, 0.01/0.00 %(u/s)time, 0.00 %gc (1x)
```


Another one
```
root@1c2a9a631f5f:/cockroach# cat cockroach-data/logs/cockroach.1c2a9a631f5f.root.2017-04-17T01_33_16Z.000001.log
I170417 01:33:16.726935 1 util/log/clog.go:1007  [config] file created at: 2017/04/17 01:33:16
I170417 01:33:16.726935 1 util/log/clog.go:1007  [config] running on machine: 1c2a9a631f5f
I170417 01:33:16.726935 1 util/log/clog.go:1007  [config] binary: CockroachDB CCL beta-20170413-dirty (linux amd64, built 2017/04/13 21:06:17, go1.8.1)
I170417 01:33:16.726935 1 util/log/clog.go:1007  [config] arguments: [/cockroach/cockroach start --insecure]
I170417 01:33:16.726935 1 util/log/clog.go:1007  line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg utf8=✓
I170417 01:33:16.726933 1 cli/start.go:390  CockroachDB CCL beta-20170413-dirty (linux amd64, built 2017/04/13 21:06:17, go1.8.1)
I170417 01:33:16.830041 20 cli/start.go:427  starting cockroach node
W170417 01:33:16.830112 20 server/server.go:158  [n?] running in insecure mode, this is strongly discouraged. See --insecure.
I170417 01:33:16.838612 20 storage/engine/rocksdb.go:379  opening rocksdb instance at "/cockroach/cockroach-data"
I170417 01:33:17.185101 20 server/config.go:403  1 storage engine initialized
I170417 01:33:17.185456 20 server/node.go:443  [n?] store [n0,s0] not bootstrapped
I170417 01:33:17.236023 20 server/node.go:378  [n?] **** cluster 2aa6cf90-c607-4d56-9d71-df535279c836 has been created
I170417 01:33:17.236036 20 server/node.go:379  [n?] **** add additional nodes by specifying --join=1c2a9a631f5f:26257
I170417 01:33:17.237796 20 storage/store.go:1256  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170417 01:33:17.237911 66 storage/split_queue.go:92  [split,n1,s1,r1/1:/M{in-ax},@c420162300] splitting at key /System/""
I170417 01:33:17.238116 20 server/node.go:456  [n1] initialized store [n1,s1]: {Capacity:63143981056 Available:56825212928 RangeCount:1 LeaseCount:1}
I170417 01:33:17.238136 20 server/node.go:347  [n1] node ID 1 initialized
I170417 01:33:17.238229 20 gossip/gossip.go:307  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"1c2a9a631f5f:26257" > attrs:<> locality:<>
I170417 01:33:17.238305 20 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
I170417 01:33:17.238420 20 server/node.go:606  [n1] connecting to gossip network to verify cluster ID...
I170417 01:33:17.238455 20 server/node.go:631  [n1] node connected via gossip and verified as part of cluster "2aa6cf90-c607-4d56-9d71-df535279c836"
I170417 01:33:17.238491 20 server/node.go:394  [n1] node=1: started with [[]=/cockroach/cockroach-data] engine(s) and attributes []
I170417 01:33:17.240998 20 sql/executor.go:326  [n1] creating distSQLPlanner with address {tcp 1c2a9a631f5f:26257}
I170417 01:33:17.245944 20 server/server.go:705  [n1] starting http server at 1c2a9a631f5f:8080
I170417 01:33:17.245958 20 server/server.go:706  [n1] starting grpc/postgres server at 1c2a9a631f5f:26257
I170417 01:33:17.245969 20 server/server.go:707  [n1] advertising CockroachDB node at 1c2a9a631f5f:26257
I170417 01:33:17.255800 20 sql/event_log.go:101  [n1] Event: "alter_table", target: 12, info: {TableName:eventlog Statement:ALTER TABLE system.eventlog ALTER COLUMN uniqueID SET DEFAULT uuid_v4() User:node MutationID:0 CascadeDroppedViews:[]}
I170417 01:33:17.258325 20 sql/lease.go:372  [n1] publish: descID=12 (eventlog) version=2 mtime=2017-04-17 01:33:17.2582982 +0000 UTC
I170417 01:33:17.264149 20 server/server.go:815  [n1] done ensuring all necessary migrations have run
I170417 01:33:17.264173 20 server/server.go:817  [n1] serving sql connections
I170417 01:33:17.264234 20 cli/start.go:497  Startup message:
CockroachDB node starting at 2017-04-17 01:33:17.2641904 +0000 UTC
build:      CCL beta-20170413-dirty @ 2017/04/13 21:06:17 (go1.8.1)
admin:      http://1c2a9a631f5f:8080
sql:        postgresql://root@1c2a9a631f5f:26257?sslmode=disable
logs:       /cockroach/cockroach-data/logs
store[0]:   path=/cockroach/cockroach-data
status:     initialized new cluster
clusterID:  2aa6cf90-c607-4d56-9d71-df535279c836
nodeID:     1
I170417 01:33:17.267728 141 sql/event_log.go:101  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:1c2a9a631f5f:26257} Attrs: Locality:} ClusterID:2aa6cf90-c607-4d56-9d71-df535279c836 StartedAt:1492392797238458100 LastUp:1492392797238458100}
I170417 01:33:17.281813 66 storage/replica_command.go:2608  [split,n1,s1,r1/1:/M{in-ax},@c420162300] initiating a split of this range at key /System/"" [r2]
I170417 01:33:17.296530 66 storage/split_queue.go:92  [split,n1,s1,r2/1:/{System/""-Max},@c420ba1500] splitting at key /System/tsd
I170417 01:33:17.296557 66 storage/replica_command.go:2608  [split,n1,s1,r2/1:/{System/""-Max},@c420ba1500] initiating a split of this range at key /System/tsd [r3]
I170417 01:33:17.300967 66 storage/split_queue.go:92  [split,n1,s1,r3/1:/{System/tsd-Max},@c420c6b200] splitting at key /System/"tse"
I170417 01:33:17.300983 66 storage/replica_command.go:2608  [split,n1,s1,r3/1:/{System/tsd-Max},@c420c6b200] initiating a split of this range at key /System/"tse" [r4]
I170417 01:33:17.307372 66 storage/split_queue.go:92  [split,n1,s1,r4/1:/{System/"tse"-Max},@c420c6bb00] splitting at key /Table/0/0
I170417 01:33:17.307388 66 storage/replica_command.go:2608  [split,n1,s1,r4/1:/{System/"tse"-Max},@c420c6bb00] initiating a split of this range at key /Table/0 [r5]
I170417 01:33:17.311026 66 storage/split_queue.go:92  [split,n1,s1,r5/1:/{Table/0-Max},@c420aa2600] splitting at key /Table/11/0
I170417 01:33:17.311040 66 storage/replica_command.go:2608  [split,n1,s1,r5/1:/{Table/0-Max},@c420aa2600] initiating a split of this range at key /Table/11 [r6]
I170417 01:33:17.314677 66 storage/split_queue.go:92  [split,n1,s1,r6/1:/{Table/11-Max},@c420cb1500] splitting at key /Table/12/0
I170417 01:33:17.314692 66 storage/replica_command.go:2608  [split,n1,s1,r6/1:/{Table/11-Max},@c420cb1500] initiating a split of this range at key /Table/12 [r7]
I170417 01:33:17.319806 66 storage/split_queue.go:92  [split,n1,s1,r7/1:/{Table/12-Max},@c420aa3b00] splitting at key /Table/13/0
I170417 01:33:17.319825 66 storage/replica_command.go:2608  [split,n1,s1,r7/1:/{Table/12-Max},@c420aa3b00] initiating a split of this range at key /Table/13 [r8]
I170417 01:33:17.323428 66 storage/split_queue.go:92  [split,n1,s1,r8/1:/{Table/13-Max},@c420988000] splitting at key /Table/14/0
I170417 01:33:17.323444 66 storage/replica_command.go:2608  [split,n1,s1,r8/1:/{Table/13-Max},@c420988000] initiating a split of this range at key /Table/14 [r9]
I170417 01:33:17.327090 66 storage/split_queue.go:92  [split,n1,s1,r9/1:/{Table/14-Max},@c420964c00] splitting at key /Table/15/0
I170417 01:33:17.327106 66 storage/replica_command.go:2608  [split,n1,s1,r9/1:/{Table/14-Max},@c420964c00] initiating a split of this range at key /Table/15 [r10]
I170417 01:33:27.239173 60 storage/store.go:3966  [n1,s1] sstables (read amplification = 0):
I170417 01:33:27.239638 62 server/status/runtime.go:227  [n1] runtime stats: 48 MiB RSS, 63 goroutines, 8.1 MiB/2.1 MiB/16 MiB GO alloc/idle/total, 3.7 MiB/6.7 MiB CGO alloc/total, 0.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (6x)
I170417 01:33:37.239336 62 server/status/runtime.go:227  [n1] runtime stats: 52 MiB RSS, 63 goroutines, 8.4 MiB/4.2 MiB/19 MiB GO alloc/idle/total, 4.1 MiB/7.2 MiB CGO alloc/total, 86.80cgo/sec, 0.01/0.00 %(u/s)time, 0.00 %gc (1x)
I170417 01:33:47.239376 62 server/status/runtime.go:227  [n1] runtime stats: 53 MiB RSS, 63 goroutines, 14 MiB/600 KiB/20 MiB GO alloc/idle/total, 4.2 MiB/7.2 MiB CGO alloc/total, 78.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (0x)
I170417 01:33:57.239443 62 server/status/runtime.go:227  [n1] runtime stats: 55 MiB RSS, 63 goroutines, 12 MiB/2.4 MiB/20 MiB GO alloc/idle/total, 4.4 MiB/7.5 MiB CGO alloc/total, 83.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (1x)
I170417 01:34:07.239437 62 server/status/runtime.go:227  [n1] runtime stats: 56 MiB RSS, 63 goroutines, 10 MiB/3.4 MiB/20 MiB GO alloc/idle/total, 4.7 MiB/7.9 MiB CGO alloc/total, 88.20cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (1x)
I170417 01:34:17.239115 61 gossip/gossip.go:452  [n1] gossip status (ok, 1 node)
gossip client (0/3 cur/max conns)
gossip server (0/3 cur/max conns, infos 0/0 sent/received, bytes 0B/0B sent/received)
I170417 01:34:17.239306 62 server/status/runtime.go:227  [n1] runtime stats: 56 MiB RSS, 63 goroutines, 9.6 MiB/4.3 MiB/20 MiB GO alloc/idle/total, 4.7 MiB/7.9 MiB CGO alloc/total, 100.90cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (1x)
I170417 01:34:27.239444 62 server/status/runtime.go:227  [n1] runtime stats: 56 MiB RSS, 63 goroutines, 8.9 MiB/4.8 MiB/20 MiB GO alloc/idle/total, 4.7 MiB/7.9 MiB CGO alloc/total, 102.60cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (1x)
I170417 01:34:30.413554 481 sql/event_log.go:101  [client=[::1]:59570,user=root,n1] Event: "create_database", target: 50, info: {DatabaseName:test Statement:CREATE DATABASE test User:root}
I170417 01:34:30.418690 481 sql/event_log.go:101  [client=[::1]:59570,user=root,n1] Event: "create_table", target: 51, info: {TableName:test Statement:CREATE TABLE test (i INT, CHECK (i > 10)) User:root}
I170417 01:34:30.420005 66 storage/split_queue.go:92  [split,n1,s1,r10/1:/{Table/15-Max},@c420bcc900] splitting at key /Table/50/0
I170417 01:34:30.420023 66 storage/replica_command.go:2608  [split,n1,s1,r10/1:/{Table/15-Max},@c420bcc900] initiating a split of this range at key /Table/50 [r11]
I170417 01:34:30.425031 66 storage/split_queue.go:92  [split,n1,s1,r11/1:/{Table/50-Max},@c420485800] splitting at key /Table/51/0
I170417 01:34:30.425047 66 storage/replica_command.go:2608  [split,n1,s1,r11/1:/{Table/50-Max},@c420485800] initiating a split of this range at key /Table/51 [r12]
panic: unsupported comparison: int to string [recovered]
        panic: create database test; set database = test; create table test (i int check (i > 10)); insert into test values ('11');: unsupported comparison: int to string [recovered]
        panic: create database test; set database = test; create table test (i int check (i > 10)); insert into test values ('11');: unsupported comparison: int to string [recovered]
        panic: create database test; set database = test; create table test (i int check (i > 10)); insert into test values ('11');: unsupported comparison: int to string

goroutine 481 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4204d0dc0)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x94
panic(0x1b29080, 0xc4201eb5e0)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc420110000, 0x7ffb836de9d0, 0xc420c480c0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:348 +0x74
panic(0x1b29080, 0xc4201eb5e0)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc42017b000, 0x74)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:487 +0x131
panic(0x1ad4700, 0xc4201eb5c0)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/parser.(*DInt).Compare(0xc42010c170, 0xc420c07040, 0x29d7bc0, 0xc420acc158, 0xc4208e3e48)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/datum.go:324 +0x1dc
github.com/cockroachdb/cockroach/pkg/sql/parser.cmpOpScalarFn(0xc420c07040, 0x29d7940, 0xc42010c170, 0x29d7bc0, 0xc420acc158, 0x1, 0x29d7900, 0x29cc320)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/eval.go:1441 +0xde
github.com/cockroachdb/cockroach/pkg/sql/parser.cmpOpScalarLTFn(0xc420c07040, 0x29d7940, 0xc42010c170, 0x29d7bc0, 0xc420acc158, 0x1, 0x29cc320, 0xc42010c170, 0x29cc420)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/eval.go:1449 +0x60
github.com/cockroachdb/cockroach/pkg/sql/parser.(*ComparisonExpr).Eval(0xc4208f4480, 0xc420c07040, 0xc4201ea190, 0x1, 0xc4201ea180, 0xc4208e3fc0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/eval.go:2409 +0x2f0
github.com/cockroachdb/cockroach/pkg/sql.(*checkHelper).check(0xc420b46f60, 0xc420c07040, 0xc420cb8140, 0x2)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/check.go:110 +0x80
github.com/cockroachdb/cockroach/pkg/sql.(*insertNode).Next(0xc420b46f20, 0x7ffb836de9d0, 0xc4202bc150, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:322 +0x253
github.com/cockroachdb/cockroach/pkg/sql.countRowsAffected(0x7ffb836de9d0, 0xc4202bc150, 0x29d57a0, 0xc420b46f20, 0xc420b46f20, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:826 +0x84
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execClassic(0xc420121040, 0xc420c07000, 0x29d57a0, 0xc420b46f20, 0xc4208e6430, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1359 +0x379
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc420121040, 0x29cc7e0, 0xc420886000, 0xc420c07000, 0xc4202bc201, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1490 +0x79d
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc420121040, 0xc42031e000, 0x29cc7e0, 0xc420886000, 0x0, 0xc420000101, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1238 +0x48d
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtsInCurrentTxn(0xc420121040, 0xc42031e000, 0xc420acc130, 0x1, 0x1, 0x0, 0x1000101, 0x0, 0xc400000000, 0xc4208e6c28, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:949 +0x818
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc420121040, 0xc42031e000, 0xc420acc130, 0x1, 0x1, 0x0, 0x1, 0xc420ba5810, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:858 +0xd1
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest.func1(0x7ffb836de9d0, 0xc4202bc150, 0xc4200c8140, 0xc420ba5810, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:690 +0x216
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc4200c8140, 0x7ffb836de9d0, 0xc4202bc150, 0x10101, 0xc420217300, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:612 +0xb6
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc420121040, 0xc42031e000, 0xc42017b000, 0x74, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc42017b000)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:712 +0x690
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc420121040, 0xc42031e000, 0xc42017b000, 0x74, 0x0, 0x0, 0x0, 0x0, 0xc42017b000)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:493 +0x11b
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc420110000, 0xc42017b000, 0x74, 0x0, 0x0, 0x0, 0x0, 0xa62a01, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:831 +0x94
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc420110000, 0xc420110028, 0xc420118240, 0x51)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:499 +0xa9
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc420110000, 0x7ffb836de9d0, 0xc420c480c0, 0xc4209e0030, 0x5400, 0xc420308310, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:427 +0x8bd
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc4203081c0, 0x7ffb836debc0, 0xc420012480, 0x29d4860, 0xc4204d0c80, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:421 +0x926
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func9.1(0x29d4860, 0xc4204d0c80)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:597 +0x153
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4204d0dc0, 0xc4201b2038, 0x29d4860, 0xc4204d0c80, 0xc4201cbd70)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
        /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x239
```

--- a/pkg/sql/check.go
+++ b/pkg/sql/check.go
@@ -72,9 +72,9 @@ func (c *checkHelper) init(
 // Set values in the IndexedVars used by the CHECK exprs.
 // Any value not passed is set to NULL, unless `merge` is true, in which
 // case it is left unchanged (allowing updating a subset of a row's values).
-func (c *checkHelper) loadRow(colIdx map[sqlbase.ColumnID]int, row parser.Datums, merge bool) {
+func (c *checkHelper) loadRow(colIdx map[sqlbase.ColumnID]int, row parser.Datums, merge bool) error {
 	if len(c.exprs) == 0 {
-		return
+		return nil
 	}
 	// Populate IndexedVars.
 	for _, ivar := range c.ivars {
@@ -83,11 +83,18 @@ func (c *checkHelper) loadRow(colIdx map[sqlbase.ColumnID]int, row parser.Datums
 		}
 		ri, has := colIdx[c.cols[ivar.Idx].ID]
 		if has {
+			if row[ri] != parser.DNull {
+				expected, provided := ivar.ResolvedType(), row[ri].ResolvedType()
+				if !expected.Equivalent(provided) {
+					return errors.Errorf("%s value does not match CHECK expr type %s", provided, expected)
+				}
+			}
 			c.curSourceRow[ivar.Idx] = row[ri]
 		} else if !merge {
 			c.curSourceRow[ivar.Idx] = parser.DNull
 		}
 	}
+	return nil
 }
 
 // IndexedVarEval implements the parser.IndexedVarContainer interface.

--- a/pkg/sql/insert.go
+++ b/pkg/sql/insert.go
@@ -318,7 +318,9 @@ func (n *insertNode) Next(ctx context.Context) (bool, error) {
 		return false, err
 	}
 
-	n.checkHelper.loadRow(n.insertColIDtoRowIndex, rowVals, false)
+	if err := n.checkHelper.loadRow(n.insertColIDtoRowIndex, rowVals, false); err != nil {
+		return false, err
+	}
 	if err := n.checkHelper.check(&n.p.evalCtx); err != nil {
 		return false, err
 	}

--- a/pkg/sql/testdata/logic_test/check_constraints
+++ b/pkg/sql/testdata/logic_test/check_constraints
@@ -5,6 +5,9 @@
 statement ok
 CREATE TABLE t1 (a INT CHECK (a > 0), to_delete INT, b INT CHECK (b < 0) CHECK (b > -100))
 
+statement error string value does not match CHECK expr type int
+INSERT INTO t1 VALUES ('3', 0, -1)
+
 statement ok
 INSERT INTO t1 VALUES (3, 0, -1)
 

--- a/pkg/sql/update.go
+++ b/pkg/sql/update.go
@@ -437,8 +437,12 @@ func (u *updateNode) Next(ctx context.Context) (bool, error) {
 		}
 	}
 
-	u.checkHelper.loadRow(u.tw.ru.FetchColIDtoRowIndex, oldValues, false)
-	u.checkHelper.loadRow(u.updateColsIdx, updateValues, true)
+	if err := u.checkHelper.loadRow(u.tw.ru.FetchColIDtoRowIndex, oldValues, false); err != nil {
+		return false, err
+	}
+	if err := u.checkHelper.loadRow(u.updateColsIdx, updateValues, true); err != nil {
+		return false, err
+	}
 	if err := u.checkHelper.check(&u.p.evalCtx); err != nil {
 		return false, err
 	}

